src/Language/PureScript/CoreFn/Desugar.hs

```diff
@@ -38,11 +38,27 @@ moduleToCoreFn env (A.Module modSS coms mn decls (Just exps)) =
   let imports = mapMaybe importToCoreFn decls ++ fmap (ssAnn modSS,) (findQualModules decls)
       imports' = dedupeImports imports
       exps' = ordNub $ concatMap exportToCoreFn exps
+      reExps = M.map ordNub $ M.unionsWith (++) (map (reExportsToCoreFn mn) exps)
       externs = ordNub $ mapMaybe externToCoreFn decls
       decls' = concatMap declToCoreFn decls
-  in Module modSS coms mn (spanName modSS) imports' exps' externs decls'
-
+  in Module modSS coms mn (spanName modSS) imports' exps' reExps externs decls'
   where
+  -- | Creates a map from a module name to the re-export references defined in
+  -- that module.
+  reExportsToCoreFn :: ModuleName -> A.DeclarationRef -> M.Map ModuleName [Ident]
+  reExportsToCoreFn mn' ref' = case ref' of
+    A.ReExportRef _ src ref ->
+      reExportsToCoreFn
+        (fromMaybe mn' (A.exportSourceImportedFrom src))
+        ref
+    _
+      | mn' == mn -> M.empty
+      | otherwise -> case ref' of
+        A.TypeRef _ _ (Just dctors) -> M.singleton mn' (fmap properToIdent dctors)
+        A.ValueRef _ name -> M.singleton mn' [name]
+        A.TypeClassRef _ name -> M.singleton mn' [properToIdent name]
+        A.TypeInstanceRef _ name -> M.singleton mn' [name]
+        _ -> M.empty
```

Comment:
Could you please add explicit cases for each DeclarationRef constructor here? That way, if we do add a new constructor or something, we'll get a warning here, so that we know to check whether this needs updating.

Reply:
Since I'm rewriting `reExportsToCoreFn` in terms of `exportToCoreFn`, I can add the explicit cases in that function. Sound good?

Reply:
Yes, that's ideal!